src/typechecker.rs

```diff
@@ -1583,6 +1583,30 @@ pub fn typecheck_namespace_predecl(
         ));
     }
 
+    // Runs a second pass through the functions to ensure that all functions get their
```

Comment:
This needs to be a pass between predecl and the full typecheck rather than something that's part of predecl. Each namespace is checked during the predecl, and we want to make sure they all get a chance to run first before we do the `typecheck_block` logic.

So, basically:

* all of predecl checks run
* then your return type inference runs
* then the rest of the typechecking runs

Reply:
So I guess it would go after `typecheck_namespace_declarations` `src/typechecker:L146`?